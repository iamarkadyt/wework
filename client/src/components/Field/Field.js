import React from 'react'
import DropdownList from 'react-widgets/lib/DropdownList'
import DateTimePicker from 'react-widgets/lib/DateTimePicker'
import Multiselect from 'react-widgets/lib/Multiselect'
import Moment from 'moment'
import momentLocalizer from 'react-widgets-moment'
import './Field.scss'
import { node, string, func, object, bool, number, arrayOf, oneOfType, instanceOf } from 'prop-types'

Moment.locale('en')
momentLocalizer()

const field = ({
    // different fields require different props
    type,
    name, value, onChange, // all except submit and button
    placeholder, error, // all except checkbox, submit and button
    label, // checkbox, submit, button
    list, // list, multiselect
    rows, // textarea
    disabled, // date
    onClick, style, // button, linkButton, textarea
    inline, // container
    containerStyle, // container
    children
}) => {
    let field = null
    const commonProps = { name, id: name, value, onChange }

    switch (type) {
        case "textarea":
            field = <textarea
                {...commonProps}
                placeholder={placeholder || 'Say something...'}
                className="rw-widget-container rw-widget-input"
                style={{
                    padding: '.5rem',
                    fontSize: '1rem',
                    width: '100%',
                    resize: 'none',
                    ...style
                }}
                rows={rows} />
            break
        case "date":
            field = <DateTimePicker
                {...commonProps}
                placeholder={placeholder || 'Pick a date...'}
                time={false}
                format="MMM DD, YYYY"
                editFormat="M/DD/YYYY"
                disabled={disabled} />
            break
        case "list":
            field = <DropdownList
                {...commonProps}
                placeholder={placeholder || "Choose one..."}
                data={list} />
            break
        case "multiselect":
            field = <Multiselect
                {...commonProps}
                placeholder={placeholder || "Enter a few items..."}
                data={list} />
            break
        case "checkbox":
            field = <label htmlFor={name}>
                <input type="checkbox" {...commonProps} />
                &nbsp;&nbsp;{label}
            </label>
            break
        case "submit":
            field = <button
                type="submit"
                className="button-action">
                {label || children}
            </button>
            break
        case "button":
            field = <button
                className="button"
                style={style}
                onClick={onClick}>
                {label || children}
            </button>
            break
        case "linkButton":
            field = <button
                className="link-button"
                style={style}
                onClick={onClick}>
                {label || children}
            </button>
            break
        default:
            field = <input
                {...commonProps}
                placeholder={placeholder}
                className="rw-widget-container rw-widget-input"
                type={type} />
    }

    return (
        <div style={containerStyle}
            className={['Field-container', inline ? 'inline' : ''].join(' ')}>
            {label
                && type !== "checkbox"
                && type !== "submit"
                && type !== "button"
                && type !== "linkButton"
                && <label htmlFor={name}>{label}</label>}
            {field}
            {error
                && type !== "checkbox"
                && type !== "submit"
                && type !== "button"
                && type !== "linkButton"
                && <span className="error">{error}</span>}
        </div>
    )
}

field.propTypes = {
  type: string,
  name: string,
  value: oneOfType([string, arrayOf(string), bool, instanceOf(Date)]),
  onChange: func,
  placeholder: string,
  error: string,
  label: string,
  list: arrayOf(string),
  rows: oneOfType([string, number]),
  disabled: bool,
  onClick: func,
  style: object,
  inline: bool,
  containerStyle: object,
  children: oneOfType([arrayOf(node), node])
}

export default field
