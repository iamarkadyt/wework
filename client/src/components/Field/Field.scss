@import 'variables.scss';
.Field-container {
    &.inline {
        display: inline;
    }
    width: 100%;
    margin: 1rem 0 0 0;
    input[type=checkbox] {
        transform: scale(1.5);
    }
    .button,
    button[type=submit] {
        min-width: 6rem;
        border-radius: .2rem;
        font-size: 1.1rem;
        padding: .2rem 1rem;
        border-color: $color-button-border;
        background: $color-button-backround;
        color: $color-button-font;
        &:active {
            border-color: $color-button-active-border;
            background: $color-button-active-background;
            color: $color-button-active-font;
        }
        &:hover:not(:active) {
            border-color: $color-button-hover-border;
            background: $color-button-hover-backround;
            color: $color-button-hover-font;
        }
    }
    button[type=submit] {
        float: right;
    }
    .button-action {
        background-color: darken(cornflowerblue, 10%);
        border-color: darken(cornflowerblue, 10%);
        background: linear-gradient(to bottom, darken(cornflowerblue, 10%), darken(cornflowerblue, 20%));
        color: white;
        &:active {
            background: linear-gradient(to bottom, darken(cornflowerblue, 25%), darken(cornflowerblue, 15%));
        }
        &:hover:not(:active) {
            background: linear-gradient(to bottom, darken(cornflowerblue, 8%), darken(cornflowerblue, 18%));
        }
    }
    .link-button {
        background: none;
        border: none;
        font-weight: bold;
        font-size: 1.1rem;
        color: $color-font-secondary;
        &:hover {
            color: $color-font-main;
            cursor: pointer;
        }
    }
    input[type=text],
    input[type=password] {
        font-size: 1rem;
        padding: .6rem;
        width: 100%;
    }
    label {
        display: inline-block;
        padding: .2rem;
    }
    textarea {
        background-color: $color-field-background;
        border-color: $color-field-border;
        color: $color-field-font;
    }
    .error {
        display: inline-block;
        color: $color-font-error;
        padding: .2rem;
    }
}