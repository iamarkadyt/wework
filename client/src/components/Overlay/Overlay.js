import React from 'react'
import './Overlay.scss'

class Overlay extends React.Component {
  componentWillMount() {
    document.body.style = "overflow: hidden"
  }

  componentWillUnmount() {
    document.body.style = null
  }

  handleSubmit(e) {
    const { 
      onBackdropClick 
    } = this.props

    if (e.target.id === 'Overlay-container-backdrop') {
      onBackdropClick()
    }
  }

  render() {
    const { 
      children, 
      centered
    } = this.props

    return (
      <div id="Overlay-container-backdrop" className="Overlay-container" onClick={this.handleSubmit}>
        <div className={["Overlay-content", centered ? "Overlay-content-centered" : null].join(' ')}>
          {children}
        </div>
      </div>
    )
  }
}

export default Overlay
