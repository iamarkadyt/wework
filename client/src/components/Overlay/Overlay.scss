@import 'variables.scss';
.Overlay-container {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    min-height: 100vh;
    display: flex;
    justify-content: center;
    align-items: center;
    padding: calc(#{$navigation_height} + 1rem) 0 2rem 0;
    .content {
        z-index: 1;
        background: white;
        padding: 0.3rem;
        border-radius: .3rem;
        box-shadow: $box_shadow_main;
    }
    .backdrop {
        position: fixed;
        top: 0;
        left: 0;
        width: 100vw;
        height: 100vh;
        background-color: rgba(0, 0, 0, 0.5);
    }
}