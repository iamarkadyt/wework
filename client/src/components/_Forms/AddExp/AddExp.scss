@mixin flexList {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;    
}

.AddExp-container {
    width: 100%;
    height: 100vh;
    @include flexList();
    form {
        @include flexList();
    }
}