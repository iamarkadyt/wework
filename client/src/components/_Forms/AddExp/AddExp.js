import React from 'react'
import { connect } from 'react-redux'
import { addExperience } from '../../../state/actions/profileActions'
import 'react-widgets/dist/css/react-widgets.css'
import './AddExp.css'

import Field from '../../Field/Field'
import Overlay from '../../Overlay/Overlay'

class AddExp extends React.Component {
    state = {}

    handleDismiss = () => {
        const { history } = this.props
        history.goBack()
    }

    render() {
        const { addExperience, errors } = this.props

        return (
            <Overlay onBackdropClick={() => this.handleDismiss()}>
                <form className="AddExp-container" onSubmit={e => {
                    e.preventDefault()
                    addExperience(this.state, () => this.handleDismiss())
                }}>
                    <h1>Add Experience</h1>
                    <Field type="text"
                        name="title"
                        placeholder="Software Developer"
                        value={this.state.title}
                        onChange={e => this.setState({ title: e.target.value })}
                        label="Title:"
                        error={errors.title} />
                    <Field type="text"
                        name="company"
                        placeholder="Facebook"
                        value={this.state.company}
                        onChange={e => this.setState({ company: e.target.value })}
                        label="Company:"
                        error={errors.company} />
                    <Field type="text"
                        name="location"
                        placeholder="San Francisco, CA"
                        value={this.state.location}
                        onChange={e => this.setState({ location: e.target.value })}
                        label="Location:"
                        error={errors.location} />
                    <Field type="date"
                        name="from"
                        value={this.state.from}
                        onChange={value => this.setState({ from: value })}
                        label="From:"
                        error={errors.from} />
                    <Field
                        type="date"
                        name="to"
                        value={this.state.to}
                        onChange={value => this.setState({ to: value })}
                        disabled={this.state.current}
                        label="To:"
                        error={errors.to} />
                    <Field
                        type="checkbox"
                        name="current"
                        value={this.state.current}
                        onChange={e => this.setState({ current: e.target.checked })}
                        label="Current?" />
                    <Field
                        type="textarea"
                        name="description"
                        value={this.state.description}
                        onChange={e => this.setState({ description: e.target.value })}
                        label="Description"
                        rows="2"
                        error={errors.description} />
                    <Field
                        type="submit"
                        label="Add" />
                    <Field
                        type="button"
                        label="Cancel"
                        onClick={e => {
                            e.preventDefault()
                            this.handleDismiss()
                        }} />
                </form>
            </Overlay>
        )
    }
}

export default connect(state => ({
    errors: state.err.formErrors
}), { addExperience })(AddExp)