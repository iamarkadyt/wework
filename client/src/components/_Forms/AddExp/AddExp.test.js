/**
 * Contract:
 *
 * What my component renders?
 * (snapshot testing)
 * - form containing 2 fields
 *   1 div with 1 submit field and 1 text span
 *
 * What do I pass to components I render?
 * (props passing)
 * Look at:
 * - form
 * - email field
 * - password field
 * - submit field
 * - Link component
 *
 * What effects do the props I receive have?
 * (props effects) > (props passing)
 * - subfields in [errors] object cause Field components to show error spans right below input fields
 * - vice versa is true: no errors, no red texts
 *
 * How do I use functions (any of them)?
 * (functions)
 * - 4 thunk functions must be called with correct args
 * - onSubmit must preventDefault(), call all 4 thunk actions and history.push()
 * - Fields' onChange functions receive event obj and call setState correctly
 *
 * What does the state hold?
 * (state) > (props passing), (interaction)
 * - email textfield data
 * - password field data
 *
 * How and When do I update/invalidate state? 
 * (state) > (interaction)
 * - When user types, respective fields update state through setState calls
 * - No invalidation occurs throughout the whole lifecycle of the component
 *
 * How state affects other components?
 * (state) > (interaction)
 * - Updates text in text Fields
 *
 * Any context?
 * (context)
 * - n/a
 *
 * Lifecycle hooks side effects?
 * - N/A
 *
 * Public ref API?
 * - N/A
 *
 * What happens when user interacts with my component?
 * (interaction)
 * - When user types, respective fields update state through setState calls
 * - When user clicks on [Submit] button form's onSubmit function is called (just ensure that type prop is right)
 * - When user clicks on signup <Link> it navigates to /signup (not your resp, just ensure props)
 */


import React from 'react'
import { _UnconnectedAddExp as AddExp } from './AddExp'
import Field from '../../Field/Field'
import { shallow } from 'enzyme'
import Overlay from '../../Overlay/Overlay'
import cloneDeep from 'lodash.clonedeep'

describe('AddExp', () => {
  let props
  let mountedAddExp

  const getMockProps = () => {
    const actions = {
      addExperience: jest.fn()
    }

    const history = {
      goBack: jest.fn()
    }

    return cloneDeep({
      ...actions,
      history,
      errors: {}
    })
  }

  const comp = () => {
    if (!mountedAddExp) {
      mountedAddExp = shallow(<AddExp {...props} />)
    }
    return mountedAddExp
  }

  beforeEach(() => {
    props = getMockProps()
    mountedAddExp = undefined
  })

  describe('props passing', () => {
    describe('Overlay', () => {
      it('receives correct set of props', () => {
        const expectedKeys = [
          'onBackdropClick',
          'children'
        ]
        expect(Object.keys(comp().find(Overlay).props())).toStrictEqual(expectedKeys)
      })
      
      it('receives correct [onBackdropClick] prop', () => {
        expect(comp().find(Overlay).prop('onBackdropClick')).toEqual(comp().instance().handleDismiss)
      })
    })

    describe('form', () => {
      it('receives correct set of props', () => {
        const expectedKeys = [
          'className',
          'onSubmit',
          'children'
        ]
        expect(Object.keys(comp().find('form').props())).toStrictEqual(expectedKeys)
      })
      
      it('receives correct [onSubmit] prop', () => {
        expect(comp().find('form').prop('onSubmit')).toEqual(expect.any(Function))
      })
    })

    describe('title Field', () => {
      it('receives correct set of props', () => {
        const expectedKeys = [
          'type',
          'name',
          'value',
          'onChange',
          'label',
          'error',
          'placeholder'
        ]
        expect(Object.keys(comp().find(Field).find('[name="title"]').props())).toStrictEqual(expectedKeys)
      })

      it('receives correct [type] prop', () => {
        expect(comp().find(Field).find('[name="title"]').prop('type')).toBe('text')
      })

      it('receives state.title into [value] prop if former is defined', () => {
        const state = 'some value'
        comp().setState({ title: state })
        expect(comp().find(Field).find('[name="title"]').prop('value')).toBe(state)
      })

      it('receives undefined into [value] prop if state.title is not defined', () => {
        comp().setState({ title: undefined })
        expect(comp().find(Field).find('[name="title"]').prop('value')).not.toBeDefined()
      })

      it('receives correct [onChange] prop', () => {
        expect(comp().find(Field).find('[name="title"]').prop('onChange')).toEqual(expect.any(Function))
      })

      it('receives correct [label] prop', () => {
        expect(comp().find(Field).find('[name="title"]').prop('label')).toBe('Title:')
      })

      it('receives props.errors.title into [error] prop if former is defined', () => {
        const msg = 'some err occured'
        props.errors.title = msg
        expect(comp().find(Field).find('[name="title"]').prop('error')).toBe(msg)
      })

      it('receives props.errors.title into [error] prop if former is defined', () => {
        delete props.errors.title
        expect(comp().find(Field).find('[name="title"]').prop('error')).not.toBeDefined()
      })

      it('receives correct [placeholder] prop', () => {
        expect(comp().find(Field).find('[name="title"]').prop('placeholder')).toBe('Software Developer')
      })
    })

    describe('company Field', () => {
      it('receives correct set of props', () => {
        const expectedKeys = [
          'type',
          'name',
          'value',
          'onChange',
          'label',
          'error',
          'placeholder'
        ]
        expect(Object.keys(comp().find(Field).find('[name="company"]').props())).toStrictEqual(expectedKeys)
      })

      it('receives correct [type] prop', () => {
        expect(comp().find(Field).find('[name="company"]').prop('type')).toBe('text')
      })

      it('receives state.company into [value] prop if former is defined', () => {
        const state = 'some value'
        comp().setState({ company: state })
        expect(comp().find(Field).find('[name="company"]').prop('value')).toBe(state)
      })

      it('receives undefined into [value] prop if state.company is not defined', () => {
        comp().setState({ company: undefined })
        expect(comp().find(Field).find('[name="company"]').prop('value')).not.toBeDefined()
      })

      it('receives correct [onChange] prop', () => {
        expect(comp().find(Field).find('[name="company"]').prop('onChange')).toEqual(expect.any(Function))
      })

      it('receives correct [label] prop', () => {
        expect(comp().find(Field).find('[name="company"]').prop('label')).toBe('Company:')
      })

      it('receives props.errors.company into [error] prop if former is defined', () => {
        const msg = 'some err occured'
        props.errors.company = msg
        expect(comp().find(Field).find('[name="company"]').prop('error')).toBe(msg)
      })

      it('receives props.errors.company into [error] prop if former is defined', () => {
        delete props.errors.company
        expect(comp().find(Field).find('[name="company"]').prop('error')).not.toBeDefined()
      })

      it('receives correct [placeholder] prop', () => {
        expect(comp().find(Field).find('[name="company"]').prop('placeholder')).toBe('Company name')
      })
    })

    describe('location Field', () => {
      it('receives correct set of props', () => {
        const expectedKeys = [
          'type',
          'name',
          'value',
          'onChange',
          'label',
          'error',
          'placeholder'
        ]
        expect(Object.keys(comp().find(Field).find('[name="location"]').props())).toStrictEqual(expectedKeys)
      })

      it('receives correct [type] prop', () => {
        expect(comp().find(Field).find('[name="location"]').prop('type')).toBe('text')
      })

      it('receives state.location into [value] prop if former is defined', () => {
        const state = 'some value'
        comp().setState({ location: state })
        expect(comp().find(Field).find('[name="location"]').prop('value')).toBe(state)
      })

      it('receives undefined into [value] prop if state.location is not defined', () => {
        comp().setState({ location: undefined })
        expect(comp().find(Field).find('[name="location"]').prop('value')).not.toBeDefined()
      })

      it('receives correct [onChange] prop', () => {
        expect(comp().find(Field).find('[name="location"]').prop('onChange')).toEqual(expect.any(Function))
      })

      it('receives correct [label] prop', () => {
        expect(comp().find(Field).find('[name="location"]').prop('label')).toBe('Location:')
      })

      it('receives props.errors.location into [error] prop if former is defined', () => {
        const msg = 'some err occured'
        props.errors.location = msg
        expect(comp().find(Field).find('[name="location"]').prop('error')).toBe(msg)
      })

      it('receives props.errors.location into [error] prop if former is defined', () => {
        delete props.errors.location
        expect(comp().find(Field).find('[name="location"]').prop('error')).not.toBeDefined()
      })

      it('receives correct [placeholder] prop', () => {
        expect(comp().find(Field).find('[name="location"]').prop('placeholder')).toBe('San Francisco, CA')
      })
    })

    describe('[from] Field', () => {
      it('receives correct set of props', () => {
        const expectedKeys = [
          'type',
          'name',
          'value',
          'onChange',
          'label',
          'error',
        ]
        expect(Object.keys(comp().find(Field).find('[name="from"]').props())).toStrictEqual(expectedKeys)
      })

      it('receives correct [type] prop', () => {
        expect(comp().find(Field).find('[name="from"]').prop('type')).toBe('date')
      })

      it('receives state.from into [value] prop if former is defined', () => {
        const state = 'some value'
        comp().setState({ from: state })
        expect(comp().find(Field).find('[name="from"]').prop('value')).toBe(state)
      })

      it('receives undefined into [value] prop if state.from is not defined', () => {
        comp().setState({ from: undefined })
        expect(comp().find(Field).find('[name="from"]').prop('value')).not.toBeDefined()
      })

      it('receives correct [onChange] prop', () => {
        expect(comp().find(Field).find('[name="from"]').prop('onChange')).toEqual(expect.any(Function))
      })

      it('receives correct [label] prop', () => {
        expect(comp().find(Field).find('[name="from"]').prop('label')).toBe('From:')
      })

      it('receives props.errors.from into [error] prop if former is defined', () => {
        const msg = 'some err occured'
        props.errors.from = msg
        expect(comp().find(Field).find('[name="from"]').prop('error')).toBe(msg)
      })

      it('receives props.errors.from into [error] prop if former is defined', () => {
        delete props.errors.from
        expect(comp().find(Field).find('[name="from"]').prop('error')).not.toBeDefined()
      })
    })

    describe('[to] Field', () => {
      it('receives correct set of props', () => {
        const expectedKeys = [
          'type',
          'name',
          'value',
          'onChange',
          'disabled',
          'label',
          'error',
        ]
        expect(Object.keys(comp().find(Field).find('[name="to"]').props())).toStrictEqual(expectedKeys)
      })

      it('receives correct [type] prop', () => {
        expect(comp().find(Field).find('[name="to"]').prop('type')).toBe('date')
      })

      it('receives state.to into [value] prop if former is defined', () => {
        const state = 'some value'
        comp().setState({ to: state })
        expect(comp().find(Field).find('[name="to"]').prop('value')).toBe(state)
      })

      it('receives undefined into [value] prop if state.to is not defined', () => {
        comp().setState({ to: undefined })
        expect(comp().find(Field).find('[name="to"]').prop('value')).not.toBeDefined()
      })

      it('receives correct [onChange] prop', () => {
        expect(comp().find(Field).find('[name="to"]').prop('onChange')).toEqual(expect.any(Function))
      })

      it('receives state.current into [disabled] prop if former is defined', () => {
        const state = false
        comp().setState({ current: state })
        expect(comp().find(Field).find('[name="to"]').prop('disabled')).toBe(state)
      })

      it('receives undefined into [disabled] prop if state.current is not defined', () => {
        comp().setState({ current: undefined })
        expect(comp().find(Field).find('[name="to"]').prop('disabled')).not.toBeDefined()
      })

      it('receives correct [label] prop', () => {
        expect(comp().find(Field).find('[name="to"]').prop('label')).toBe('To:')
      })

      it('receives props.errors.to into [error] prop if former is defined', () => {
        const msg = 'some err occured'
        props.errors.to = msg
        expect(comp().find(Field).find('[name="to"]').prop('error')).toBe(msg)
      })

      it('receives props.errors.to into [error] prop if former is defined', () => {
        delete props.errors.to
        expect(comp().find(Field).find('[name="to"]').prop('error')).not.toBeDefined()
      })
    })

    describe('[current] Field', () => {
      it('receives correct set of props', () => {
        const expectedKeys = [
          'type',
          'name',
          'value',
          'onChange',
          'label',
        ]
        expect(Object.keys(comp().find(Field).find('[name="current"]').props())).toStrictEqual(expectedKeys)
      })

      it('receives correct [type] prop', () => {
        expect(comp().find(Field).find('[name="current"]').prop('type')).toBe('checkbox')
      })

      it('receives correct [label] prop', () => {
        expect(comp().find(Field).find('[name="current"]').prop('label')).toBe('Current?')
      })

      it('receives correct [onChange] prop', () => {
        expect(comp().find(Field).find('[name="current"]').prop('onChange')).toEqual(expect.any(Function))
      })

      it('receives state.current into [value] prop if former is defined', () => {
        const state = true
        comp().setState({ current: state })
        expect(comp().find(Field).find('[name="current"]').prop('value')).toBe(state)
      })

      it('receives undefined into [value] prop if state.current is not defined', () => {
        comp().setState({ current: undefined })
        expect(comp().find(Field).find('[name="current"]').prop('value')).not.toBeDefined()
      })
    })

    describe('description Field', () => {
      it('receives correct set of props', () => {
        const expectedKeys = [
          'type',
          'name',
          'value',
          'onChange',
          'label',
          'rows',
          'error',
        ]
        expect(Object.keys(comp().find(Field).find('[name="description"]').props())).toStrictEqual(expectedKeys)
      })

      it('receives correct [type] prop', () => {
        expect(comp().find(Field).find('[name="description"]').prop('type')).toBe('textarea')
      })

      it('receives state.description into [value] prop if former is defined', () => {
        const state = 'some value'
        comp().setState({ description: state })
        expect(comp().find(Field).find('[name="description"]').prop('value')).toBe(state)
      })

      it('receives undefined into [value] prop if state.description is not defined', () => {
        comp().setState({ description: undefined })
        expect(comp().find(Field).find('[name="description"]').prop('value')).not.toBeDefined()
      })

      it('receives correct [onChange] prop', () => {
        expect(comp().find(Field).find('[name="description"]').prop('onChange')).toEqual(expect.any(Function))
      })

      it('receives correct [label] prop', () => {
        expect(comp().find(Field).find('[name="description"]').prop('label')).toBe('Description')
      })

      it('receives props.errors.description into [error] prop if former is defined', () => {
        const msg = 'some err occured'
        props.errors.description = msg
        expect(comp().find(Field).find('[name="description"]').prop('error')).toBe(msg)
      })

      it('receives props.errors.description into [error] prop if former is defined', () => {
        delete props.errors.description
        expect(comp().find(Field).find('[name="description"]').prop('error')).not.toBeDefined()
      })

      it('receives correct [rows] prop', () => {
        expect(comp().find(Field).find('[name="description"]').prop('rows')).toBe('2')
      })
    })

    describe('submit button', () => {
      it('receives correct set of props', () => {
        const expectedKeys = [
          'type',
          'name',
          'label',
        ]
        expect(Object.keys(comp().find(Field).find('[name="submit"]').props())).toStrictEqual(expectedKeys)
      })

      it('receives correct [type] prop', () => {
        expect(comp().find(Field).find('[name="submit"]').prop('type')).toBe('submit')
      })

      it('receives correct [label] prop', () => {
        expect(comp().find(Field).find('[name="submit"]').prop('label')).toBe('Add')
      })
    })

    describe('cancel button', () => {
      it('receives correct set of props', () => {
        const expectedKeys = [
          'type',
          'name',
          'label',
          'onClick'
        ]
        expect(Object.keys(comp().find(Field).find('[name="cancel"]').props())).toStrictEqual(expectedKeys)
      })

      it('receives correct [type] prop', () => {
        expect(comp().find(Field).find('[name="cancel"]').prop('type')).toBe('button')
      })

      it('receives correct [label] prop', () => {
        expect(comp().find(Field).find('[name="cancel"]').prop('label')).toBe('Cancel')
      })

      it('receives correct [onClick] prop', () => {
        expect(comp().find(Field).find('[name="cancel"]').prop('onClick')).toEqual(expect.any(Function))
      })
    })
  })

  describe('interaction', () => {
    describe('form submission', () => {
      let mockedEvent

      beforeEach(() => {
        mockedEvent = { preventDefault: jest.fn() }
      })

      it('invokes preventDefault()', () => {
        comp().find('form').simulate('submit', mockedEvent)
        expect(mockedEvent.preventDefault).toHaveBeenCalled()
      })

      it('invokes addExperience()', () => {
        comp().find('form').simulate('submit', mockedEvent)
        expect(props.addExperience).toHaveBeenCalled()
      })

      it('invokes addExperience() with correct args', () => {
        comp().find('form').simulate('submit', mockedEvent)
        expect(props.addExperience).toHaveBeenCalledWith(comp().state(), comp().instance().handleDismiss)
      })
    })

    describe('Fields', () => {
      let mockedEvent
      let value = 'some value'

      beforeEach(() => {
        mockedEvent = { 
          target: { value }, 
          preventDefault: jest.fn()
        }
      })
      
      describe('title', () => {
        it('[onChange] function works as expected', () => {
          comp().find(Field).find('[name="title"]').simulate('change', mockedEvent)
          expect(comp().state('title')).toBe(value)
        })
      })

      describe('company', () => {
        it('[onChange] function works as expected', () => {
          comp().find(Field).find('[name="company"]').simulate('change', mockedEvent)
          expect(comp().state('company')).toBe(value)
        })
      })

      describe('location', () => {
        it('[onChange] function works as expected', () => {
          comp().find(Field).find('[name="location"]').simulate('change', mockedEvent)
          expect(comp().state('location')).toBe(value)
        })
      })

      describe('to', () => {
        it('[onChange] function works as expected', () => {
          comp().find(Field).find('[name="to"]').simulate('change', value)
          expect(comp().state('to')).toBe(value)
        })
      })

      describe('from', () => {
        it('[onChange] function works as expected', () => {
          comp().find(Field).find('[name="from"]').simulate('change', value)
          expect(comp().state('from')).toBe(value)
        })
      })

      describe('description', () => {
        it('[onChange] function works as expected', () => {
          comp().find(Field).find('[name="description"]').simulate('change', mockedEvent)
          expect(comp().state('description')).toBe(value)
        })
      })

      describe('cancel', () => {
        describe('[onClick]', () => {
          it('invokes preventDefault()', () => {
            comp().find(Field).find('[name="cancel"]').simulate('click', mockedEvent)
            expect(mockedEvent.preventDefault).toHaveBeenCalled()
          })

          it('invokes handleDismiss()', () => {
            const mockFn = jest.fn()
            comp().instance().handleDismiss = mockFn
            comp().find(Field).find('[name="cancel"]').simulate('click', mockedEvent)
            expect(mockFn).toHaveBeenCalled()
          })
        })
      })
    })
  })

  it('matches snapshot', () => {
    expect(comp()).toMatchSnapshot()
  })
})
