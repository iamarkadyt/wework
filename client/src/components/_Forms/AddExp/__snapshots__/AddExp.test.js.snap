// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddExp matches snapshot 1`] = `
<Overlay
  onBackdropClick={[Function]}
>
  <form
    className="AddExp-container"
    onSubmit={[Function]}
  >
    <h1>
      Add Experience
    </h1>
    <field
      label="Title:"
      name="title"
      onChange={[Function]}
      placeholder="Software Developer"
      type="text"
    />
    <field
      label="Company:"
      name="company"
      onChange={[Function]}
      placeholder="Company name"
      type="text"
    />
    <field
      label="Location:"
      name="location"
      onChange={[Function]}
      placeholder="San Francisco, CA"
      type="text"
    />
    <field
      label="From:"
      name="from"
      onChange={[Function]}
      type="date"
    />
    <field
      label="To:"
      name="to"
      onChange={[Function]}
      type="date"
    />
    <field
      label="Current?"
      name="current"
      onChange={[Function]}
      type="checkbox"
    />
    <field
      label="Description"
      name="description"
      onChange={[Function]}
      rows="2"
      type="textarea"
    />
    <field
      label="Add"
      name="submit"
      type="submit"
    />
    <field
      label="Cancel"
      name="cancel"
      onClick={[Function]}
      type="button"
    />
  </form>
</Overlay>
`;
